src/CommandLine.cc

```diff
@@ -0,0 +1,445 @@
+#if USE_WIN32_SERVICE
+            "       -O options\n"
+            "                 Set Windows Service Command line options in Registry.\n"
+#endif
+            "       -R        Do not set REUSEADDR on port.\n"
+            "       -S        Double-check swap during rebuild.\n"
+            "       -X        Force full debugging.\n"
+            "       -Y        Only return UDP_HIT or UDP_MISS_NOFETCH during fast reload.\n",
+            APP_SHORTNAME, CACHE_HTTP_PORT, DefaultConfigFile, CACHE_ICP_PORT);
+    exit(EXIT_FAILURE);
+}
+
+CommandLine::CommandLine(const int anArgc, char *anArgv[])
+    : execFile_(anArgv[0])
+{
+    parse(anArgc, anArgv);
+}
+
+void
+CommandLine::parse(int anArgc, char *anArgv[])
+{
+    int optIndex = 0;
+
+    // short options
+    const char *shortOpStr =
+#if USE_WIN32_SERVICE
+        "O:Vir"
+#endif
+        "CDFNRSYXa:d:f:hk:m::n:sl:u:vz?";
+
+    // long options
+    static struct option squidOptions[] = {
+        {"foreground", no_argument, 0,  ForegroundCode},
+        {"kid",        required_argument, 0, KidCode},
+        {"help",       no_argument, 0, 'h'},
+        {"version",    no_argument, 0, 'v'},
+        {0, 0, 0, 0}
+    };
+
+    int c;
+
+    while ((c = getopt_long(anArgc, anArgv, shortOpStr, squidOptions, &optIndex)) != -1) {
+        if (c == '?' || c == 'h') {
+            usage();
+            // unreacheable
+            break;
+        }
+        options.push_back(std::make_pair(c, optarg ? SBuf(optarg) : SBuf()));
+    }
+
+    // reserve space for --kid option (with argument) and nil termination pointer
+    argv_.clear();
+    argv_.reserve(anArgc + 3);
+    for (int i = 0; i < anArgc; ++i)
+        argv_.push_back(anArgv[i]);
+}
+
+const char **
+CommandLine::argv(const char *argv0, const char *kid)
+{
+    argv_[0] = argv0;
+
+    auto kidPos = std::find_if(argv_.begin(), argv_.end(), [](const char *opt)
+            { return strcmp(opt, "--kid") == 0; });
+    // not expected to happen because kids do not create kids
+    if (kidPos != argv_.end()) {
+        assert(++kidPos != argv_.end());
+        *kidPos = xstrdup(kid);
+    } else {
+        argv_.push_back("--kid");
+        argv_.push_back(xstrdup(kid));
+    }
+    argv_.push_back(nullptr);
+    return &argv_[0];
+}
+
+SBuf
+CommandLine::kidName() const
+{
+    auto kidOpt = std::find_if(options.begin(), options.end(), [](const OptionsPair &p)
+             { return p.first == KidCode; });
+    if (kidOpt == options.end())
+        return SBuf();
+    return kidOpt->second;
+}
+
+void
+CommandLine::processOptions()
+{
+    std::for_each(options.begin(), options.end(), [this](OptionsPair &opt)
+            { processOption(opt.first, opt.second.isEmpty() ? nullptr : opt.second.c_str()); });
+}
+
+// apply a single option
+void
+CommandLine::processOption(const char optCode, const char *optArg)
+{
+    // XXX: use optArg instead
+    optarg = const_cast<char *>(optArg);
+    switch (optCode) {
```

Comment:
Please move this `switch` statement back into main.cc so that CommandLine knows nothing (or at least a lot less) about ./squid-specific options?

Reply:
We can add a function pointer parameter to CommandLine::processOptions() and pass this (moved to main.cc) method to it.

Reply:
Yes, a callback sounds like the best option to me (at least among the simple ones).